File: samples/Windows/java/TableManagerSamples/CloseAllPositions/src/TableListener.java
package closeallpositions;

import java.util.ArrayList;
import java.util.List;

import com.fxcore2.*;
import common.BatchOrderMonitor;
import common.OrderMonitor;

public class TableListener implements IO2GTableListener {
    private ResponseListener mResponseListener;
    private BatchOrderMonitor mBatchOrderMonitor;
    private List<String> mRequestIDs;

    // ctor
    public TableListener(ResponseListener responseListener) {
        mResponseListener = responseListener;
        mBatchOrderMonitor = null;
        mRequestIDs = new ArrayList<String>();
    }

    public void setRequestIDs(List<String> requestIDs) {
        mRequestIDs.clear();
        for (String sRequestID : requestIDs) {
            mRequestIDs.add(sRequestID);
        }
        mBatchOrderMonitor = new BatchOrderMonitor();
        mBatchOrderMonitor.setRequestIDs(mRequestIDs);
    }

    public void onAdded(String sRowID, O2GRow rowData) {
        O2GTableType type = rowData.getTableType();
        switch (type) {
        case ORDERS:
            O2GOrderRow orderRow = (O2GOrderRow)rowData;
            if (mBatchOrderMonitor != null && mRequestIDs.contains(orderRow.getRequestID())) {
                System.out.println(String.format("The order has been added. Order ID: %s, Rate: %s, Time In Force: %s",
                        orderRow.getOrderID(),
                        orderRow.getRate(),
                        orderRow.getTimeInForce()));
                mBatchOrderMonitor.onOrderAdded(orderRow);
            }
            break;
        case TRADES:
            O2GTradeRow tradeRow = (O2GTradeRow)rowData;
            if (mBatchOrderMonitor != null) {
                mBatchOrderMonitor.onTradeAdded(tradeRow);
                if (mBatchOrderMonitor.isBatchExecuted()) {
                    printResult();
                    mResponseListener.stopWaiting();
                }
            }
            break;
        case CLOSED_TRADES:
            O2GClosedTradeRow closedTradeRow = (O2GClosedTradeRow)rowData;
            if (mBatchOrderMonitor != null) {
                mBatchOrderMonitor.onClosedTradeAdded(closedTradeRow);
                if (mBatchOrderMonitor.isBatchExecuted()) {
                    printResult();
                    mResponseListener.stopWaiting();
                }
            }
            break;
        case MESSAGES:
            O2GMessageRow messageRow = (O2GMessageRow)rowData;
            if (mBatchOrderMonitor != null) {
                mBatchOrderMonitor.onMessageAdded(messageRow);
                if (mBatchOrderMonitor.isBatchExecuted()) {
                    printResult();
                    mResponseListener.stopWaiting();
                }
            }
            break;
        }
    }

    public void onChanged(String sRowID, O2GRow rowData) {
        if (rowData.getTableType() == O2GTableType.ACCOUNTS) {
            O2GAccountTableRow account = (O2GAccountTableRow)rowData;
            System.out.println(String.format("Balance: %.2f, Equity: %.2f", account.getBalance(), account.getEquity()));
        }
    }

    public void onDeleted(String sRowID, O2GRow rowData) {
        if (rowData.getTableType() == O2GTableType.ORDERS) {
            O2GOrderRow orderRow = (O2GOrderRow)rowData;
            if (mRequestIDs.contains(orderRow.getRequestID())) {
                if (mBatchOrderMonitor != null) {
                    System.out.println(String.format("The order has been deleted. Order ID: %s", orderRow.getOrderID()));
                    mBatchOrderMonitor.onOrderDeleted(orderRow);
                    if (mBatchOrderMonitor.isBatchExecuted()) {
                        printResult();
                        mResponseListener.stopWaiting();
                    }
                }
            }
        }
    }

    public void onStatusChanged(O2GTableStatus status) {
    }

    private void printResult() {
        for (OrderMonitor monitor : mBatchOrderMonitor.getMonitors()) {
            System.out.println(String.format("Result for OrderID %s:", monitor.getOrder().getOrderID()));
            printMonitorResult(monitor);
        }
    }

    private void printMonitorResult(OrderMonitor orderMonitor) {
        if (orderMonitor != null) {
            OrderMonitor.ExecutionResult result = orderMonitor.getResult();
            List<O2GTradeRow> trades;
            List<O2GClosedTradeRow> closedTrades;
            O2GOrderRow order = orderMonitor.getOrder();
            String sOrderID = order.getOrderID();
            trades = orderMonitor.getTrades();
            closedTrades = orderMonitor.getClosedTrades();

            switch (result) {
            case Canceled:
                if (trades.size() > 0) {
                    printTrades(trades, sOrderID);
                    printClosedTrades(closedTrades, sOrderID);
                    System.out.println(String.format("A part of the order has been canceled. Amount = %s",
                            orderMonitor.getRejectAmount()));
                } else {
                    System.out.println(String.format(
                            "The order: OrderID = %s  has been canceled.",
                            sOrderID));
                    System.out.println(String.format(
                            "The cancel amount = %s.",
                            orderMonitor.getRejectAmount()));
                }
                break;
            case FullyRejected:
                System.out.println(String.format(
                        "The order has been rejected. OrderID = %s",
                        sOrderID));
                System.out.println(String.format("The rejected amount = %s",
                        orderMonitor.getRejectAmount()));
                System.out.println(String.format("Rejection cause: %s",
                        orderMonitor.getRejectMessage()));
                break;
            case PartialRejected:
                printTrades(trades, sOrderID);
                printClosedTrades(closedTrades, sOrderID);
                System.out.println(String.format(
                        "A part of the order has been rejected. Amount = %s",
                        orderMonitor.getRejectAmount()));
                System.out.println(String.format("Rejection cause: %s ",
                        orderMonitor.getRejectMessage()));
                break;
            case Executed:
                printTrades(trades, sOrderID);
                printClosedTrades(closedTrades, sOrderID);
                break;
            }
        }
    }

    private void printTrades(List<O2GTradeRow> trades, String sOrderID) {
        if (trades.size() == 0)
            return;
        System.out.println(String.format("For the order: OrderID = %s the following positions have been opened:", sOrderID));

        for (int i = 0; i < trades.size(); i++) {
            O2GTradeRow trade = trades.get(i);
            String sTradeID = trade.getTradeID();
            int iAmount = trade.getAmount();
            double dRate = trade.getOpenRate();
            System.out.println(String.format("Trade ID: %s; Amount: %s; Rate: %s", sTradeID, iAmount, dRate));
        }
    }

    private void printClosedTrades(List<O2GClosedTradeRow> closedTrades, String sOrderID) {
        if (closedTrades.size() == 0)
            return;
        System.out.println(String.format("For the order: OrderID = %s the following positions have been closed: ", sOrderID));

        for (int i = 0; i < closedTrades.size(); i++) {
            O2GClosedTradeRow closedTrade = closedTrades.get(i);
            String sTradeID = closedTrade.getTradeID();
            int iAmount = closedTrade.getAmount();
            double dRate = closedTrade.getCloseRate();
            System.out.println(String.format("Closed Trade ID: %s; Amount: %s; Closed Rate: %s", sTradeID, iAmount, dRate));
        }
    }

    public void subscribeEvents(O2GTableManager manager) {
        O2GAccountsTable accountsTable = (O2GAccountsTable)manager.getTable(O2GTableType.ACCOUNTS);
        O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.ORDERS);
        O2GTradesTable tradesTable = (O2GTradesTable)manager.getTable(O2GTableType.TRADES);
        O2GMessagesTable messagesTable = (O2GMessagesTable)manager.getTable(O2GTableType.MESSAGES);
        O2GClosedTradesTable closedTradesTable = (O2GClosedTradesTable)manager.getTable(O2GTableType.CLOSED_TRADES);
        accountsTable.subscribeUpdate(O2GTableUpdateType.UPDATE, this);
        ordersTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
        ordersTable.subscribeUpdate(O2GTableUpdateType.DELETE, this);
        tradesTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
        tradesTable.subscribeUpdate(O2GTableUpdateType.UPDATE, this);
        closedTradesTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
        messagesTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
    }

    public void unsubscribeEvents(O2GTableManager manager) {
        O2GAccountsTable accountsTable = (O2GAccountsTable)manager.getTable(O2GTableType.ACCOUNTS);
        O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.ORDERS);
        O2GTradesTable tradesTable = (O2GTradesTable)manager.getTable(O2GTableType.TRADES);
        O2GMessagesTable messagesTable = (O2GMessagesTable)manager.getTable(O2GTableType.MESSAGES);
        O2GClosedTradesTable closedTradesTable = (O2GClosedTradesTable)manager.getTable(O2GTableType.CLOSED_TRADES);
        accountsTable.unsubscribeUpdate(O2GTableUpdateType.UPDATE, this);
        ordersTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
        ordersTable.unsubscribeUpdate(O2GTableUpdateType.DELETE, this);
        tradesTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
        tradesTable.unsubscribeUpdate(O2GTableUpdateType.UPDATE, this);
        closedTradesTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
        messagesTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
    }
}

File: samples/Linux/netcore_for_Linux_x64_only/netcore1.0/TableManagerSamples/CloseAllPositions/src/TableListener.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text;
using fxcore2;
using OrdersMonitor;

namespace CloseAllPositions
{
    class TableListener : IO2GTableListener
    {
        private ResponseListener mResponseListener;
        private BatchOrderMonitor mBatchOrderMonitor;
        private List<string> mRequestIDs;

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="responseListener"></param>
        public TableListener(ResponseListener responseListener)
        {
            mResponseListener = responseListener;
            mBatchOrderMonitor = null;
            mRequestIDs = new List<string>();
        }

        public void SetRequestIDs(List<string> requestIDs)
        {
            mRequestIDs.Clear();
            foreach (string sRequestID in requestIDs)
            {
                mRequestIDs.Add(sRequestID);
            }
            mBatchOrderMonitor = new BatchOrderMonitor();
            mBatchOrderMonitor.SetRequestIDs(mRequestIDs);
        }

        #region IO2GTableListener Members

        public void onAdded(string sRowID, O2GRow rowData)
        {
            O2GTableType type = rowData.TableType;
            switch (type)
            {
                case O2GTableType.Orders:
                    O2GOrderRow orderRow = (O2GOrderRow)rowData;
                    if (mBatchOrderMonitor != null && mRequestIDs.Contains(orderRow.RequestID))
                    {
                        Console.WriteLine("The order has been added. Order ID: {0}, Rate: {1}, Time In Force: {2}",
                            orderRow.OrderID,
                            orderRow.Rate,
                            orderRow.TimeInForce);
                        mBatchOrderMonitor.OnOrderAdded(orderRow);
                    }
                    break;
                case O2GTableType.Trades:
                    O2GTradeRow tradeRow = (O2GTradeRow)rowData;
                    if (mBatchOrderMonitor != null)
                    {
                        mBatchOrderMonitor.OnTradeAdded(tradeRow);
                        if (mBatchOrderMonitor.IsBatchExecuted())
                        {
                            PrintResult();
                            mResponseListener.StopWaiting();
                        }
                    }
                    break;
                case O2GTableType.ClosedTrades:
                    O2GClosedTradeRow closedTradeRow = (O2GClosedTradeRow)rowData;
                    if (mBatchOrderMonitor != null)
                    {
                        mBatchOrderMonitor.OnClosedTradeAdded(closedTradeRow);
                        if (mBatchOrderMonitor.IsBatchExecuted())
                        {
                            PrintResult();
                            mResponseListener.StopWaiting();
                        }
                    }
                    break;
                case O2GTableType.Messages:
                    O2GMessageRow messageRow = (O2GMessageRow)rowData;
                    if (mBatchOrderMonitor != null)
                    {
                        mBatchOrderMonitor.OnMessageAdded(messageRow);
                        if (mBatchOrderMonitor.IsBatchExecuted())
                        {
                            PrintResult();
                            mResponseListener.StopWaiting();
                        }
                    }
                    break;
            }
        }

        public void onChanged(string sRowID, O2GRow rowData)
        {
            if (rowData.TableType == O2GTableType.Accounts)
            {
                O2GAccountTableRow account = (O2GAccountTableRow)rowData;
                Console.WriteLine("Balance: {0}, Equity: {1}", account.Balance, account.Equity);
            }
        }

        public void onDeleted(string sRowID, O2GRow rowData)
        {
            if (rowData.TableType == O2GTableType.Orders)
            {
                O2GOrderRow orderRow = (O2GOrderRow)rowData;
                if (mRequestIDs.Contains(orderRow.RequestID))
                {
                    Console.WriteLine("The order has been deleted. Order ID: {0}", orderRow.OrderID);
                    if (mBatchOrderMonitor != null)
                    {
                        mBatchOrderMonitor.OnOrderDeleted(orderRow);
                        if (mBatchOrderMonitor.IsBatchExecuted())
                        {
                            PrintResult();
                            mResponseListener.StopWaiting();
                        }
                    }
                }
            }
        }

        public void onStatusChanged(O2GTableStatus status)
        {
        }

        #endregion

        private void PrintResult()
        {
            foreach (OrderMonitor monitor in mBatchOrderMonitor.GetMonitors())
            {
                Console.WriteLine("Result for OrderID {0}:", monitor.Order.OrderID);
                PrintMonitorResult(monitor);
            }
        }

        private void PrintMonitorResult(OrderMonitor orderMonitor)
        {
            if (orderMonitor != null)
            {
                OrderMonitor.ExecutionResult result = orderMonitor.Result;
                ReadOnlyCollection<O2GTradeRow> trades;
                ReadOnlyCollection<O2GClosedTradeRow> closedTrades;
                O2GOrderRow order = orderMonitor.Order;
                String orderID = order.OrderID;
                trades = orderMonitor.Trades;
                closedTrades = orderMonitor.ClosedTrades;

                switch (result)
                {
                    case OrderMonitor.ExecutionResult.Canceled:
                        if (trades.Count > 0)
                        {
                            PrintTrades(trades, orderID);
                            PrintClosedTrades(closedTrades, orderID);
                            Console.WriteLine("A part of the order has been canceled. Amount = {0}", orderMonitor.RejectAmount);
                        }
                        else
                        {
                            Console.WriteLine("The order: OrderID = {0}  has been canceled.", orderID);
                            Console.WriteLine("The cancel amount = {0}.", orderMonitor.RejectAmount);
                        }
                        break;
                    case OrderMonitor.ExecutionResult.FullyRejected:
                        Console.WriteLine("The order has been rejected. OrderID = {0}", orderID);
                        Console.WriteLine("The rejected amount = {0}", orderMonitor.RejectAmount);
                        Console.WriteLine("Rejection cause: {0}", orderMonitor.RejectMessage);
                        break;
                    case OrderMonitor.ExecutionResult.PartialRejected:
                        PrintTrades(trades, orderID);
                        PrintClosedTrades(closedTrades, orderID);
                        Console.WriteLine("A part of the order has been rejected. Amount = {0}", orderMonitor.RejectAmount);
                        Console.WriteLine("Rejection cause: {0} ", orderMonitor.RejectMessage);
                        break;
                    case OrderMonitor.ExecutionResult.Executed:
                        PrintTrades(trades, orderID);
                        PrintClosedTrades(closedTrades, orderID);
                        break;
                }
            }
        }

        private void PrintTrades(ReadOnlyCollection<O2GTradeRow> trades, string orderID)
        {
            if (trades.Count == 0)
                return;
            Console.WriteLine("For the order: OrderID = {0} the following positions have been opened:", orderID);

            for (int i = 0; i < trades.Count; i++)
            {
                O2GTradeRow trade = trades[i];
                String tradeID = trade.TradeID;
                int amount = trade.Amount;
                double rate = trade.OpenRate;
                Console.WriteLine("Trade ID: {0}; Amount: {1}; Rate: {2}", tradeID, amount, rate);
            }
        }

        private void PrintClosedTrades(ReadOnlyCollection<O2GClosedTradeRow> closedTrades, string orderID)
        {
            if (closedTrades.Count == 0)
                return;
            Console.WriteLine("For the order: OrderID = {0} the following positions have been closed: ", orderID);

            for (int i = 0; i < closedTrades.Count; i++)
            {
                O2GClosedTradeRow closedTrade = closedTrades[i];
                String tradeID = closedTrade.TradeID;
                int amount = closedTrade.Amount;
                double rate = closedTrade.CloseRate;
                Console.WriteLine("Closed Trade ID: {0}; Amount: {1}; Closed Rate: {2}", tradeID, amount, rate);
            }
        }

        public void SubscribeEvents(O2GTableManager manager)
        {
            O2GAccountsTable accountsTable = (O2GAccountsTable)manager.getTable(O2GTableType.Accounts);
            O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.Orders);
            O2GTradesTable tradesTable = (O2GTradesTable)manager.getTable(O2GTableType.Trades);
            O2GMessagesTable messagesTable = (O2GMessagesTable)manager.getTable(O2GTableType.Messages);
            O2GClosedTradesTable closedTradesTable = (O2GClosedTradesTable)manager.getTable(O2GTableType.ClosedTrades);
            accountsTable.subscribeUpdate(O2GTableUpdateType.Update, this);
            ordersTable.subscribeUpdate(O2GTableUpdateType.Insert, this);
            ordersTable.subscribeUpdate(O2GTableUpdateType.Delete, this);
            tradesTable.subscribeUpdate(O2GTableUpdateType.Insert, this);
            tradesTable.subscribeUpdate(O2GTableUpdateType.Update, this);
            closedTradesTable.subscribeUpdate(O2GTableUpdateType.Insert, this);
            messagesTable.subscribeUpdate(O2GTableUpdateType.Insert, this);
        }

        public void UnsubscribeEvents(O2GTableManager manager)
        {
            O2GAccountsTable accountsTable = (O2GAccountsTable)manager.getTable(O2GTableType.Accounts);
            O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.Orders);
            O2GTradesTable tradesTable = (O2GTradesTable)manager.getTable(O2GTableType.Trades);
            O2GMessagesTable messagesTable = (O2GMessagesTable)manager.getTable(O2GTableType.Messages);
            O2GClosedTradesTable closedTradesTable = (O2GClosedTradesTable)manager.getTable(O2GTableType.ClosedTrades);
            accountsTable.unsubscribeUpdate(O2GTableUpdateType.Update, this);
            ordersTable.unsubscribeUpdate(O2GTableUpdateType.Insert, this);
            ordersTable.unsubscribeUpdate(O2GTableUpdateType.Delete, this);
            tradesTable.unsubscribeUpdate(O2GTableUpdateType.Insert, this);
            tradesTable.unsubscribeUpdate(O2GTableUpdateType.Update, this);
            closedTradesTable.unsubscribeUpdate(O2GTableUpdateType.Insert, this);
            messagesTable.unsubscribeUpdate(O2GTableUpdateType.Insert, this);
        }

    }
}
